app/models/concerns/smart_proxy_host_extensions.rb

```diff
@@ -7,8 +7,12 @@ def smart_proxy_reference(hash)
       ProxyReferenceRegistry.add_smart_proxy_reference hash
     end
 
+    def smart_proxy_scope(hosts_scope = Host::Managed.all)
```

Comment:
Nitpick: we can make it a scope:

```suggestion
    scope :with_smart_proxies, -> { eager_load(ProxyReferenceRegistry.smart_proxy_references.select(&:join?).map(&:join_relation) }
```

I have dereferenced the `proxy_join_tables` method. I suppose we can reuse it, if the method will be defined as static.

then later we can use it directly:

``` ruby
hosts_scope.with_smart_proxies.pluck(proxy_column_list).flatten.uniq.compact
```